### ISSUE TITLE
escapeJavaScript ecaping in request templates

### ISSUE THREAD

**Issue by russau** [aws/serverless-application-model#1895]
https://github.com/aws/serverless-application-model/blob/16fa8522d596a153965d6801943249276b8d974b/samtranslator/model/stepfunctions/events.py#L437

Using `escapeJavaScript` here can cause problems with apostrophes.  More details in the forum posts:

https://forums.aws.amazon.com/thread.jspa?threadID=228917
https://forums.aws.amazon.com/thread.jspa?messageID=710397

**Comment by hoffa** [comment]
Thanks for reporting this!

Do you have a template that would help reproducing any issue you've had?

**Comment by russau** [comment]
Hello @hoffa, I have a template below that will reproduce what I'm seeing.

I see the mapping template generated looks like

``` json
{"input": "$util.escapeJavaScript($input.json('$'))", "stateMachineArn": "arn:aws:states:us-west-2:..."}
```

A post to my `echo` API without any single quotes works:

```
curl --location --request POST 'https://....execute-api.us-west-2.amazonaws.com/Prod/echo' \
--header 'Content-Type: application/json' \
--data-raw '{"hello": "world"}'
```

A post that contains quotes fails with `{"__type":"com.amazon.coral.service#SerializationException"}`.

```
curl --location --request POST 'https://....execute-api.us-west-2.amazonaws.com/Prod/echo' \
--header 'Content-Type: application/json' \
--data-raw '{"hello": "'\''world'\''"}'
```

The template..

``` yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31
Description: Repro

Resources:
  MyApi:
    Type: AWS::Serverless::Api
    Properties:
      StageName: Prod

  HelloWorldFunction:
    Type: AWS::Serverless::Function
    Properties:
      InlineCode: |
        def handler(event, context):
            print(event)
            return "do nothing"
      Handler: index.handler
      Runtime: python3.8


  Post:
    Type: AWS::Serverless::StateMachine
    Properties:
      Role: null
      Policies:
        - arn:aws:iam::aws:policy/AWSLambdaFullAccess
      Definition:
        StartAt: One
        States:
          One:
            Type: Task
            Resource: !GetAtt HelloWorldFunction.Arn
            End: true
      Events:
        PostEcho:
          Type: Api
          Properties:
            RestApiId: !Ref MyApi
            Path: /echo
            Method: POST
```

**Comment by hoffa** [comment]
I'm able to reproduce, thanks for making it so easy!

**Comment by russau** [comment]
There's an example in the [$util](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html#util-template-reference) template documentation about replacing apostrophes to make everything json compatible:

`$util.escapeJavaScript(data).replaceAll("\\'","'")`

**Comment by danemauland** [comment]
Has there been any traction on this? I just ran into this issue, and I'm very surprised that SAM isn't designed to handle such a common use case.

**Comment by hoffa** [comment]
Thanks for your patience, I have opened a PR for this.

**Comment by SuttonO-Webeo** [comment]
> Thanks for your patience, I have opened PR #2253 for this.

Any updates on when this PR might get merged?

**Comment by hoffa** [comment]
https://github.com/aws/serverless-application-model/pull/2591 is merged; it adds a new `UnescapeMappingTemplate` boolean property to state machine `Api` events. It will roll out over the coming weeks.

For context, the reason for this being an opt-in property (and partly why we were a little reluctant to merge the initial PRs) is to avoid large-scale redeployments due to the logical ID changes (see https://github.com/aws/serverless-application-model/pull/2039).

**Comment by aahung** [comment]
Released in [v1.56.0](https://github.com/aws/serverless-application-model/releases/tag/v1.56.0)
